evmc/evmc.nim

```diff
@@ -271,7 +271,7 @@ type
   # The result is passed by pointer to avoid (shallow) copy of the ::evmc_result
   # struct. Think of this as the best possible C language approximation to
   # passing objects by reference.
-  evmc_release_result_fn* = proc(result: ptr evmc_result) {.cdecl.}
+  evmc_release_result_fn* = proc(result: var evmc_result) {.cdecl.}
```

Comment:
aren't these part of the "low-level" C interface?

Reply:
It is, but the two "high-level" interfaces are the implementation procs and the caller procs (layers upon layers), which already use `var` almost everywhere.

As it says in the commit/PR intro:

> Use `var` instead of `ptr` in the Nim EVMC API `import evmc/evmc` because **all users of the API are already using `var`**

> API users **use casts to ignore type mismatches**. Therefore, change the API itself to use `var`, so it's better to use directly, casts are not needed, and proper type checking can be used.

The question of whether to write a `{.cdecl.}` C pointer argument as `ptr` or `var` doesn't have a settled answer as far as I 'm aware. I have seen both used elsewhere. Thus the list of four ways to write a C pointer argument:

> we can represent function arguments which are pointers in C as any of these in Nim: [...]

I think it's not really useful to stick with `ptr` in these signatures, unless we want the signatures to carry on serving just as human documentation, if it means everything that uses this module is motivated to ignore the type signatures entirely by using function pointer casts. Currently they don't really need most of the contents of this file because the types are ignored, and they are using their own type definitons.

So the idea is to go from `evmc.nim` being a mixture of useful object types and mere-documentation proc signatures, to real typed proc signatures that the higher layers actually use.

Of course it's possible to change those other places to use `ptr`, but then yet another layer would be required to translate to what they actually need in the function body. Is it useful or pragmatic? The reason current implementation procs use `var` without explicit translation procs in between the two styles is because pragmatically we all know `var` arguments are pointers, and they are ok to use with `{.cdecl.}`.

Finally I don't think `ptr` even serves well to document low-level C argument types, because most of the C types are _const_ pointers (and the constness is really important in this API, they point to the internals of client data structures that must never by changed), and `ptr` doesn't say that.

Whereas Nim does represent that in a canonical way: The Nim way to describe pointer-to-mutable and pointer-to-const is generally `var arg` and just `arg`, as long as the latter will be passed by reference.

Thus, along with some people on some forums somewhere, I'm taking the view that `var` is a perfectly legitimate way to document a C-compatible function argument of type "pointer to ...".